# Contents
01. Import Libraries
02. Importing Datasets
03. Lesson
04. 4.7 Exercise
05. Export as Pickle

# 01. Import Libraries

In [16]:
import pandas as pd
import numpy as np

In [17]:
# Confirming that pandas & numpy were imported by printing versions

print(pd.__version__)
print(np.__version__)

2.1.4
1.26.4


# 02. Importing Datasets

In [19]:
# Defines path as intro file path for Instacart project

path = r'/home/0668a905-109d-4403-be59-0a04abf51dd9/Instacart Basket Analysis'

In [21]:
# Imports orders_products_variables dataframe

ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_variables.pkl'))

In [23]:
# Create subset dataframe "df" to reduce size

df = ords_prods_merge[:1000000]

In [25]:
# Check shape

df.shape

(1000000, 21)

In [27]:
# sort by user_id so that the following exercises will be more clear

df = df.sort_values('user_id')

In [29]:
# Check output

df.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
470349            17   3194192        2             7                   2   
274200            11   2168274        2             1                   2   
267310            12   1501582        2             2                   5   
534079            16   1199898        2             6                   2   
545095            13   1901567        2             3                   1   

        order_hour_of_day  days_since_prior_order  product_id  \
470349                 12                    14.0       24852   
274200                 11                     NaN       47766   
267310                 10                    10.0       47766   
534079                  9                    13.0       24852   
545095                 10                     3.0        7781   

        add_to_cart_order  reordered  ...  \
470349                  8          1  ...   
274200                  2          0  ...   
267310                  4          1  ...   
534079                  7          1  ...   
545095                  4          0  ...   

                                             product_name aisle_id  \
470349                                             Banana       24   
274200                                    Organic Avocado       24   
267310                                    Organic Avocado       24   
534079                                             Banana       24   
545095  Organic Sticks Low Moisture Part Skim Mozzarel...       21   

        department_id  prices  _merge    price_range_loc     busiest_day  \
470349              4    12.3    both  Mid-range product  Regularly busy   
274200              4     6.3    both  Mid-range product  Regularly busy   
267310              4     6.3    both  Mid-range product  Regularly busy   
534079              4    12.3    both  Mid-range product  Regularly busy   
545095             16     6.6    both  Mid-range product  Regularly busy   

          busiest_days busiest_period_of_day busiest_period_of_day_2  
470349  Regularly busy           Most orders             Most orders  
274200  Regularly busy           Most orders             Most orders  
267310  Regularly busy           Most orders             Most orders  
534079  Regularly busy           Most orders             Most orders  
545095    Busiest days           Most orders             Most orders  

[5 rows x 21 columns]

# 03. Lesson

In [31]:
# Group df by column 'product_name'

df.groupby('product_name')

In [33]:
# Grouping by department_id and aggregating by the average ord_number

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                16.378579
3                17.689479
4                18.040567
5                16.243119
6                14.021088
7                16.853578
9                17.687095
11               16.370763
12               16.488347
13               16.272080
14               17.702703
15               17.657026
16               18.216617
17               17.099583
18               18.627069
19               17.965219
20               15.556348

In [35]:
# Grouping by dept id and gathering mean of order_number using .mean instead of .agg()

df.groupby('department_id')['order_number'].mean()

department_id
1     16.378579
3     17.689479
4     18.040567
5     16.243119
6     14.021088
7     16.853578
9     17.687095
11    16.370763
12    16.488347
13    16.272080
14    17.702703
15    17.657026
16    18.216617
17    17.099583
18    18.627069
19    17.965219
20    15.556348
Name: order_number, dtype: float64

In [37]:
# Using aggregate function to find multiple stats (mean, min, max) of order # grouped by dept id

df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                16.378579   1  99
3                17.689479   1  99
4                18.040567   1  99
5                16.243119   1  99
6                14.021088   1  96
7                16.853578   1  99
9                17.687095   1  99
11               16.370763   1  96
12               16.488347   1  99
13               16.272080   1  99
14               17.702703   1  99
15               17.657026   1  99
16               18.216617   1  99
17               17.099583   1  99
18               18.627069   1  99
19               17.965219   1  99
20               15.556348   1  99

In [39]:
# Create a loyalty flag using .loc

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/tmp/ipykernel_2572/3640534792.py:3: FutureWarning: The provided callable <function max at 0x7f4cf11dd360> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [41]:
# Tells panda to not put a max on rows displayed

pd.options.display.max_rows = None

In [43]:
ords_prods_merge.head(100)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0         927121   2617132    55630             6                   2   
1        2792928     30342   168460            31                   2   
2        2193387   3191294   132038             4                   6   
3         700457    747623    42201            28                   0   
4         214674     68597    12960            19                   6   
5         459480   1629767    27656             6                   4   
6        1735969   3056006   104237             6                   3   
7         613000    431664    36965            29                   5   
8        1156351   1231608    69505            18                   6   
9        1910433   2766802   114711            64                   2   
10        315030   3365670    19061             2                   0   
11       1235430   2789502    74277            14                   1   
12       1305725   1686685    78478             6                   6   
13       3361707   1781309   202623            96                   3   
14       2577918    463539   155135            26                   4   
15        250545   1204803    15155            13                   6   
16       2678847    690540   161323             8                   1   
17        686788     35513    41388            22                   1   
18       1134919   2113968    68238             8                   5   
19       1177325   1646835    70763            13                   1   
20       3118122    435434   188012             3                   0   
21        452795   1410344    27262            26                   2   
22       2058710   2818956   123751            27                   6   
23       1317424   3280138    79186             5                   6   
24       2910529   1595057   175591            11                   1   
25        395667    932101    23769             7                   1   
26       2926738   1168498   176576            27                   6   
27       1075075   2988290    64775            16                   3   
28        582954   1618882    35107            30                   6   
29       1589997   1945476    95476            12                   4   
30       1085986   2646491    65415             4                   2   
31       3241098   3228466   195356            29                   6   
32       1846994   1532109   110865             3                   5   
33       2830965    265632   170798            13                   4   
34       2302353   2175320   138637            11                   0   
35       2100258   2082664   126281            29                   2   
36       1265438   1562043    76106            53                   1   
37       2659717   2473695   160099             7                   6   
38       1963499    843672   117917            33                   4   
39       1967329   1261229   118147            13                   5   
40        667209   1052950    40197            14                   3   
41        312246   1984310    18878             8                   1   
42       1846332   1386106   110833            16                   3   
43       3048143   2373765   183894            13                   5   
44       1371911   3070596    82437            21                   5   
45        442445   2226421    26626            37                   1   
46          9768   2159933      618            14                   3   
47       1739214   1255475   104417            28                   1   
48       1805743     87518   108452             6                   3   
49        217190      8943    13120            15                   3   
50       2615529   2458148   157482            24                   0   
51       1157770    220968    69584             2                   4   
52        419647   1825880    25261             4                   1   
53       1429965   2600238    85928            40             

In [45]:
# Create flag for 'loyal customer'

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [47]:
# Create flag for 'Regular customer'

ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [49]:
# Create flag for 'New customer'

ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [51]:
# Check frequency of customer flags

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    3155832
Loyal customer      2046053
New customer        1279059
Name: count, dtype: int64

In [53]:
# Check output for specific columns

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0     55630    Loyal customer             6
1    168460    Loyal customer            31
2    132038      New customer             4
3     42201    Loyal customer            28
4     12960  Regular customer            19
5     27656  Regular customer             6
6    104237  Regular customer             6
7     36965  Regular customer            29
8     69505  Regular customer            18
9    114711    Loyal customer            64
10    19061      New customer             2
11    74277    Loyal customer            14
12    78478  Regular customer             6
13   202623    Loyal customer            96
14   155135  Regular customer            26
15    15155  Regular customer            13
16   161323  Regular customer             8
17    41388  Regular customer            22
18    68238    Loyal customer             8
19    70763  Regular customer            13
20   188012    Loyal customer             3
21    27262  Regular customer            26
22   123751    Loyal customer            27
23    79186      New customer             5
24   175591    Loyal customer            11
25    23769  Regular customer             7
26   176576    Loyal customer            27
27    64775    Loyal customer            16
28    35107  Regular customer            30
29    95476  Regular customer            12
30    65415  Regular customer             4
31   195356  Regular customer            29
32   110865  Regular customer             3
33   170798  Regular customer            13
34   138637  Regular customer            11
35   126281  Regular customer            29
36    76106    Loyal customer            53
37   160099  Regular customer             7
38   117917  Regular customer            33
39   118147  Regular customer            13
40    40197  Regular customer            14
41    18878  Regular customer             8
42   110833    Loyal customer            16
43   183894  Regular customer            13
44    82437  Regular customer            21
45    26626    Loyal customer            37
46      618  Regular customer            14
47   104417  Regular customer            28
48   108452      New customer             6
49    13120  Regular customer            15
50   157482    Loyal customer            24
51    69584  Regular customer             2
52    25261      New customer             4
53    85928  Regular customer            40
54    30783      New customer             2
55   197413  Regular customer            10
56   160999    Loyal customer            48
57   186736  Regular customer             9
58    80686      New customer             2
59   115413      New customer             3

# 04. 4.7 Exercise

## Find aggregated mean of "order_number" for entire data frame (not just subset df)

In [55]:
# Grouping by department_id and aggregating by the average ord_number for entire dataset

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.483696
2                17.402979
3                17.152759
4                17.815772
5                15.306562
6                16.409440
7                17.271845
8                15.152595
9                15.904668
10               19.985087
11               16.216951
12               15.936314
13               16.590358
14               16.778791
15               16.194325
16               17.656715
17               15.710896
18               19.237574
19               17.224106
20               16.463765
21               22.946905

#### Difference between order mean by subset compared to entire dataframe

Dept 1 in subset had an order number mean of 16.378579 and the same dept in entire df had a mean of 15.483696.

Dept 21 in subset had an order number mean of 15.556348 and the same dept in entire df had a mean of 22.946905.

Incorporating more data changed the averages taking into account that the first subset was limited to the first 1,000,000 rows. In a sense they become more extreme (smaller and larger averages).


### Price differences between Loyal, Regular, and New customers

In [57]:
# Using aggregate function to find multiple stats (mean, min, max) of order # grouped by customer loyalty

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.443980  1.0  99999.0
New customer      14.232985  1.0  99999.0
Regular customer  12.023942  1.0  99999.0

In [59]:
# Using aggregate function to find multiple stats (mean, min, max) of order # grouped by customer loyalty

ords_prods_merge.groupby('loyalty_flag').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                         mean min max
loyalty_flag                         
Loyal customer      32.734885   1  99
New customer         3.843439   1  10
Regular customer    12.446532   1  40

#### Notes when comparing basic stats of prices vs order number

The mean price of the loyal customers appears to be lower than the new & regular customers. New customers appear to be paying higher prices on average. However, when you look at the order numbers, loyal customers of course have a much higher average count than the new and regular customers.


## Low & High Spender Tags

In [61]:
# Create a Spender Range column using transformation aggregation that gathers average price spent by customer

ords_prods_merge['mean_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/tmp/ipykernel_2572/1472052255.py:3: FutureWarning: The provided callable <function mean at 0x7f4cf11ddc60> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['mean_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [63]:
# Check output for specific columns

ords_prods_merge[['user_id', 'prices', 'mean_price']].head()

user_id  prices  mean_price
0    55630    14.3    7.550000
1   168460    14.3    7.696932
2   132038    14.3    9.258333
3    42201    14.3    7.596947
4    12960    14.3    7.863953

In [65]:
# Create flag for 'High spender'

ords_prods_merge.loc[ords_prods_merge['mean_price'] >= 10, 'spender_flag'] = 'High spender'

In [67]:
# Create flag for 'Low spender'

ords_prods_merge.loc[ords_prods_merge['mean_price'] < 10, 'spender_flag'] = 'Low spender'

In [69]:
# Check frequency of spender flags

ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low spender     6322937
High spender     158007
Name: count, dtype: int64

## Customer Frequency Tag

In [71]:
# Create a Median Order column using transformation aggregation that gathers median order made by customer

ords_prods_merge['median_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/tmp/ipykernel_2572/3001223843.py:3: FutureWarning: The provided callable <function median at 0x7f4cf0dd0d30> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [73]:
# Check output for specific columns

ords_prods_merge[['user_id', 'days_since_prior_order', 'median_order']].head()

user_id  days_since_prior_order  median_order
0    55630                     5.0           5.0
1   168460                     6.0           7.0
2   132038                    30.0          13.0
3    42201                     2.0           5.0
4    12960                     6.0           8.0

In [75]:
# Create flag for 'Non-frequent customer'

ords_prods_merge.loc[ords_prods_merge['median_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [77]:
# Create flag for 'Regular customer'

ords_prods_merge.loc[(ords_prods_merge['median_order'] > 10) & (ords_prods_merge['median_order'] <=20), 'order_frequency_flag'] = 'Regular customer'

In [79]:
# Create flag for 'Frequent customer'

ords_prods_merge.loc[ords_prods_merge['median_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [81]:
# Check frequency of order frequency flags

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        4306996
Regular customer         1416195
Non-frequent customer     754546
nan                         3207
Name: count, dtype: int64

In [83]:
# Check output for specific columns

ords_prods_merge[['user_id', 'days_since_prior_order', 'median_order', 'order_frequency_flag']].head()

user_id  days_since_prior_order  median_order order_frequency_flag
0    55630                     5.0           5.0    Frequent customer
1   168460                     6.0           7.0    Frequent customer
2   132038                    30.0          13.0     Regular customer
3    42201                     2.0           5.0    Frequent customer
4    12960                     6.0           8.0    Frequent customer

In [85]:
print(ords_prods_merge[ords_prods_merge['order_frequency_flag'] == 'nan'][['user_id', 'days_since_prior_order', 'median_order', 'order_frequency_flag']].head())

       user_id  days_since_prior_order  median_order order_frequency_flag
1810      3574                     NaN           NaN                  nan
6987     52245                     NaN           NaN                  nan
13131   102973                     NaN           NaN                  nan
13758   190082                     NaN           NaN                  nan
17950   134217                     NaN           NaN                  nan


#### Order Frequency Notes:

There were 3207 rows with missing information on customer's days since prior order. This could possibly mean that they've made only one order from Instacart. In that case, the NaN values should have been treated as "0" and be considered non-frequent customers.

# 05. Export as Pickle

In [133]:
# Drop unnecessary columns that need removed before merging DFs

ords_prods_merge_2 = ords_prods_merge.drop(columns = ['Unnamed: 0_x', 'Unnamed: 0_y', '_merge', 'busiest_period_of_day_2'])

In [136]:
# Set the display option to show all columns

pd.set_option('display.max_columns', None)

In [139]:
# Double check output before exporting 

ords_prods_merge_2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2617132    55630             6                   2                 15   
1     30342   168460            31                   2                 12   
2   3191294   132038             4                   6                 12   
3    747623    42201            28                   0                 10   
4     68597    12960            19                   6                  9   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     5.0       31766                 13          0   
1                     6.0       31766                 13          1   
2                    30.0       31766                  4          0   
3                     2.0       31766                 10          1   
4                     6.0       31766                 16          1   

          product_name  aisle_id  department_id  prices    price_range_loc  \
0  Cheese Pizza Snacks       129              1    14.3  Mid-range product   
1  Cheese Pizza Snacks       129              1    14.3  Mid-range product   
2  Cheese Pizza Snacks       129              1    14.3  Mid-range product   
3  Cheese Pizza Snacks       129              1    14.3  Mid-range product   
4  Cheese Pizza Snacks       129              1    14.3  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy           Most orders         50   
1  Regularly busy  Regularly busy           Most orders         43   
2  Regularly busy  Regularly busy           Most orders          9   
3     Busiest day    Busiest days           Most orders         62   
4  Regularly busy  Regularly busy           Most orders         22   

       loyalty_flag  mean_price spender_flag  median_order  \
0    Loyal customer    7.550000  Low spender           5.0   
1    Loyal customer    7.696932  Low spender           7.0   
2      New customer    9.258333  Low spender          13.0   
3    Loyal customer    7.596947  Low spender           5.0   
4  Regular customer    7.863953  Low spender           8.0   

  order_frequency_flag  
0    Frequent customer  
1    Frequent customer  
2     Regular customer  
3    Frequent customer  
4    Frequent customer

In [142]:
# Export as pickle

ords_prods_merge_2.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_aggregates.pkl'))